### PR TITLE
[doc] Document the floating-point and bitvector primitives

### PR DESCRIPTION
**Pull request by bclement-ocp**
Nominating for 2.5.0 @Halbaroth I guess :)

### CODE REVIEW

File: docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Alt-Ergo has partial support for the [SMT-LIB
+standard](http://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf)
+language from the SMT community.
+
+*Note*: As of version 2.5.0, enhanced support for the SMT-LIB language is
+provided by the new [Dolmen](http://gbury.github.io/dolmen/) frontend. To use
+it, pass the option `--frontend dolmen` to Alt-Ergo. This will become the
+default in a future release.
+
```
Comment (Halbaroth) [inline docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md:12]: I wad doing the same doc...

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md [hunk 2]
```diff
@@ -129,7 +127,73 @@ goal g:
...
+### Conversion between integers and reals
+
+```alt-ergo
+logic real_of_int : int -> real
+```
+
+`real_of_int` converts an integer into its representation as a real number.
+
+*Note*: When using the `dolmen` frontend, `real_of_int` is also available in
+the smtlib2 format as the `to_real` function from the `Reals_Ints` theory.
```
Comment (Halbaroth) [inline docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md:145]: Maybe a table given the translation between these AE primitives and the SMT-LIB ones would be clearer?

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md [hunk 3]
```diff
@@ -129,7 +127,73 @@ goal g:
...
+
+```alt-ergo
+logic int_floor : real -> int
+logic int_ceil : real -> int
+logic real_is_int : real -> bool
+```
+
+`int_floor` and `int_ceil` implement the standard `floor` and `ceil` functions.
+They compute the greatest integer less than a real and the least integer
+greater than a real, respectively.
```
Comment (Halbaroth) [inline docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: By standard, you mean the SMT-LIB standard or the mathematical one? I guess they agree but who knows :)

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md [hunk 4]
```diff
@@ -129,7 +127,73 @@ goal g:
...
+from the `Reals_Ints` theory respectively.
+
+### Square root
+
+```alt-ergo
+logic sqrt_real : real -> real
+```
+
+The `sqrt_real` function denotes the square root of a real number.  Alt-Ergo
+only computes rational under- and over-approximations.
```
Comment (Halbaroth) [inline docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: Does it mean `sqrt_real` computes the under-approximation or the over-approximation?

File: src/bin/text/index.mld [hunk 5]
```diff
@@ -5,9 +5,11 @@
 The alt-ergo package installs the {e alt-ergo} binary, whose documentation
 is available through the {e --help} option.
 
+See also the {{:https://ocamlpro.github.io/alt-ergo/}language documentation}.
+
```
Comment (Halbaroth) [inline src/bin/text/index.mld:9]: Why do you add a link here? There is very same link on the sidebar.

File: docs/sphinx_docs/Input_file_formats/Native/05_theories.md [hunk 6]
```diff
@@ -37,12 +37,77 @@ All theories are always considered *modulo equality*.
...
 
+These functions are currently only available when using the native language;
+they are not available when using the smtlib2 input format.
+
+Finally, the `integer_round` function allows rounding a real to an integer
+using the aforementioned rounding modes:
+
+```alt-ergo
+logic integer_round : fpa_rounding_mode, real -> int
+```
```
Comment (Halbaroth) [inline docs/sphinx_docs/Input_file_formats/Native/05_theories.md:110]: You could add an example of input file using the round function? Or a link to an examples in the appropriate directory of the repository.

**Review by Halbaroth — APPROVED** [review]

File: docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md (hunk 1 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md:12]: The BV primitives are only available using the Dolmen frontend, I had to mention it. I can remove that paragraph.

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md (hunk 4 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: `sqrt_real` computes the square root. Internally, Alt-Ergo tracks both an under- and an over- approximation of the square root as rationals. I agree that this sentence is confusing, I will remove it.

File: src/bin/text/index.mld (hunk 5 above)
Comment (bclement-ocp) [reply src/bin/text/index.mld:9]: This is the odoc for the alt-ergo package (this page: https://ocamlpro.github.io/alt-ergo/odoc/dev/alt-ergo/index.html ), I don't see a link to the language doc in the sidebar?

(Sorry, this probably should have been in a separate PR, I just did it as I was in the doc folder)

File: docs/sphinx_docs/Input_file_formats/Native/05_theories.md (hunk 6 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/Native/05_theories.md:110]: Sure.

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md (hunk 3 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: There are no `floor` or `ceil` functions in the SMT-LIB standard. Changed to usual instead.

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md (hunk 2 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md:145]: I don't think a table would be very helpful, people usually use either AE or SMT-LIB. This is in the documentation for the native input language anyways, so I'll remove these notes and add corresponding paragraphs in the SMT-LIB format doc.

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md (hunk 4 above)
Comment (Halbaroth) [reply docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: But I agree we shouldn't write `the square root` without mentioning it's an approximation.

File: src/bin/text/index.mld (hunk 5 above)
Comment (Halbaroth) [reply src/bin/text/index.mld:9]: Oh sorry, I thought it was in the sphinx documentation. You don't need to do a separate PR for such a small modification.

File: docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md (hunk 1 above)
Comment (Halbaroth) [reply docs/sphinx_docs/Input_file_formats/SMT-LIB2/index.md:12]: Let's keep your paragraph ;)

File: docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md (hunk 4 above)
Comment (bclement-ocp) [reply docs/sphinx_docs/Input_file_formats/Native/02_types/02_01_builtin.md]: It is not an approximation.

Edited to clarify: if `e` is not a constant rational square (i.e. equal to `r * r` for some rational `r`), Alt-Ergo treats `sqrt_real(e)` as uninterpreted, with the constraint that `sqrt_real_default(e) ≤ sqrt_real(e) ≤ sqrt_real_excess(e)` where `sqrt_real_default` and `sqrt_real_excess` are the under- and over-approximation. `sqrt_real_default` and `sqrt_real_excess` always compute to rational numbers for all (constant) rational inputs, but `sqrt_real` doesn't.
(Also in all cases we do have the constraint `sqrt_real(x) * sqrt_real(x) = x` of course)

**Pull request merged**